packages/hardhat-core/src/internal/hardhat-network/provider/TxPool.ts

```diff
@@ -0,0 +1,432 @@
+
+  private readonly _deserializeTransaction: (
+    tx: SerializedTransaction
+  ) => OrderedTransaction;
+
+  constructor(
+    private readonly _stateManager: PStateManager,
+    blockGasLimit: BN,
+    common: Common
+  ) {
+    this._state = makePoolState({
+      blockGasLimit: bnToHex(blockGasLimit),
+    });
+    this._deserializeTransaction = (tx) => deserializeTransaction(tx, common);
+  }
+
+  public async addTransaction(tx: Transaction) {
+    const senderNonce = await this._validateTransaction(tx);
+    const txNonce = new BN(tx.nonce);
+
+    if (txNonce.eq(senderNonce)) {
+      this._addPendingTransaction(tx);
+    } else {
+      this._addQueuedTransaction(tx);
+    }
+  }
+
+  public snapshot(): number {
+    if (this._snapshotIdToState.get(this._currentSnapshotId) !== this._state) {
+      this._currentSnapshotId = this._nextSnapshotId++;
+      this._snapshotIdToState.set(this._currentSnapshotId, this._state);
+    }
+    return this._currentSnapshotId;
+  }
+
+  public revert(snapshotId: number) {
+    const state = this._snapshotIdToState.get(snapshotId);
+    if (state === undefined) {
+      throw new Error("There's no snapshot with such ID");
+    }
+    this._currentSnapshotId = snapshotId;
+    this._state = state;
+  }
+
+  public getTransactionByHash(hash: Buffer): OrderedTransaction | undefined {
+    const tx = this._getTransactionsByHash().get(bufferToHex(hash));
+    if (tx !== undefined) {
+      return this._deserializeTransaction(tx);
+    }
+
+    return undefined;
+  }
+
+  public hasPendingTransactions(): boolean {
+    const pendingMap = this._getPending();
+    return pendingMap.some((senderPendingTxs) => !senderPendingTxs.isEmpty());
+  }
+
+  public getPendingTransactions(): Map<string, OrderedTransaction[]> {
+    const deserializedImmutableMap = this._getPending()
+      .filter((txs) => txs.size > 0)
+      .map((txs) => txs.map(this._deserializeTransaction).toJS());
+    return new Map(deserializedImmutableMap.entries());
+  }
+
+  public getQueuedTransactions(): Map<string, OrderedTransaction[]> {
+    const deserializedImmutableMap = this._getQueued()
+      .filter((txs) => txs.size > 0)
+      .map((txs) => txs.map(this._deserializeTransaction).toJS());
+    return new Map(deserializedImmutableMap.entries());
+  }
+
+  public async getExecutableNonce(accountAddress: Buffer): Promise<BN> {
+    const nonce = this._getExecutableNonces().get(bufferToHex(accountAddress));
+    if (nonce === undefined) {
+      const account = await this._stateManager.getAccount(accountAddress);
+      return new BN(account.nonce);
+    }
+    return new BN(toBuffer(nonce));
+  }
+
+  public getBlockGasLimit(): BN {
+    return new BN(toBuffer(this._state.get("blockGasLimit")));
+  }
+
+  public setBlockGasLimit(newLimit: BN | number) {
+    if (typeof newLimit === "number") {
+      newLimit = new BN(newLimit);
+    }
+
+    this._setBlockGasLimit(newLimit);
+  }
+
+  public async clean() {
```

Comment:
There's a subtle bug here. It goes like this:

1. Disable automine and interval mining
2. Send a transaction that has a gas limit of, say, 100K
3. Set the block gas limit to a lower value, like 80K
4. Send a new transaction with the same account

The network basically stops working after this. The reason is that the transaction in step 2 is dropped, and the one sent in step 4 has the wrong nonce (one more than the one in 2), but the node tries to execute it anyway, and the VM fails. In other words, the transaction in step 4 should be queued but it's added to pending.

I'm not sure about the fix, I guess the `executableNonces` need to be updated after cleaning here?